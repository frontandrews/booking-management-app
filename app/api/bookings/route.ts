import { NextResponse, NextRequest } from 'next/server';
import axios from 'axios';
import { API_URL } from '@/constants';
import { isValid, parseISO } from 'date-fns';
import { handleAxiosError } from '@/utils';

export async function POST(request: NextRequest) {
  try {
    const { name, propertyId, startDate, endDate, pricePerDay } =
      await request.json();

    if (!name || !propertyId || !startDate || !endDate || !pricePerDay) {
      return NextResponse.json(
        {
          message:
            'name, propertyId, startDate, endDate, and pricePerDay are required',
        },
        { status: 400 },
      );
    }

    if (!isValid(parseISO(startDate)) || !isValid(parseISO(endDate))) {
      return NextResponse.json(
        { message: 'Invalid date format' },
        { status: 400 },
      );
    }

    const response = await axios.post(`${API_URL}/bookings`, {
      name,
      propertyId,
      startDate,
      endDate,
      pricePerDay,
    });
    return NextResponse.json(response.data, { status: 201 });
  } catch (error: any) {
    return handleAxiosError(error);
  }
}

export async function GET() {
  try {
    const response = await axios.get(`${API_URL}/bookings`);
    return NextResponse.json(response.data, { status: 200 });
  } catch (error: any) {
    return handleAxiosError(error);
  }
}
