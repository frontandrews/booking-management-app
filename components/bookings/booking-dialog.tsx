import { useState, ReactNode } from 'react';
import { Button } from '@/components/ui/button';
import {
  Dialog,
  DialogContent,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog';
import BookingForm from './booking-form';
import { Plus } from 'lucide-react';

interface BookingDialogProps {
  bookingId?: number;
  buttonClassName?: string;
  children?: ReactNode;
}

export function BookingDialog({
  bookingId,
  buttonClassName,
  children,
}: BookingDialogProps) {
  const [isOpen, setIsOpen] = useState(false);
  const handleClose = () => setIsOpen(false);

  return (
    <Dialog open={isOpen} onOpenChange={setIsOpen}>
      <DialogTrigger asChild>
        {children ? (
          <div className={buttonClassName} onClick={() => setIsOpen(true)}>
            {children}
          </div>
        ) : (
          <Button
            variant="default"
            size="sm"
            className={`ml-2 ${buttonClassName}`}
            onClick={() => setIsOpen(true)}
          >
            <Plus className="w-4 h-4 mr-2" />
            Reservation
          </Button>
        )}
      </DialogTrigger>
      <DialogContent className="sm:max-w-[425px] pb-6">
        <DialogHeader className="px-6 pt-4">
          <DialogTitle>
            {bookingId ? 'Edit Reservation' : 'New Reservation'}
          </DialogTitle>
        </DialogHeader>
        <BookingForm bookingId={bookingId} onClose={handleClose} />
      </DialogContent>
    </Dialog>
  );
}
