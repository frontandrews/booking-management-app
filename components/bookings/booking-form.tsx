'use client';

import { zodResolver } from '@hookform/resolvers/zod';
import { useForm, FormProvider } from 'react-hook-form';
import { z } from 'zod';
import { useToast } from '@/components/ui/use-toast';
import {
  createBooking,
  updateBooking,
  deleteBooking,
  fetchBookings,
  fetchBookingById,
} from '@/redux/features/booking/slice';
import { Button } from '@/components/ui/button';
import { useEffect, useState } from 'react';
import { getErrorMessage } from '@/utils';
import { parseISO, startOfDay, format } from 'date-fns';
import { DatePickerWithRange } from '@/components/ui/date-picker-with-range';
import { ThreeDots } from 'react-loader-spinner';
import { TrashIcon } from 'lucide-react';
import FormField from '@/components/ui/form-field';
import { useAppDispatch, useAppSelector } from '@/redux/hooks';

const formSchema = z.object({
  name: z.string().min(1, 'Name is required'),
  propertyId: z.number().min(1, 'Property ID is required'),
  dateRange: z
    .object({
      from: z.date().nullable(),
      to: z.date().nullable(),
    })
    .refine(
      (data) => data.from && data.to,
      'Both start and end dates are required',
    ),
  pricePerDay: z.string().min(1, 'Price per day is required'),
});

interface BookingFormProps {
  bookingId?: number;
  onClose: () => void;
}

export default function BookingForm({ bookingId, onClose }: BookingFormProps) {
  const { toast } = useToast();
  const [formError, setFormError] = useState<string>('');
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [totalPrice, setTotalPrice] = useState<number>(0);
  const dispatch = useAppDispatch();
  const bookings = useAppSelector((state) => state.booking.bookings);
  const loading = useAppSelector((state) => state.booking.loading);
  const propertyId = useAppSelector(
    (state) => state.property.selectedPropertyId,
  );

  const methods = useForm({
    resolver: zodResolver(formSchema),
    defaultValues: {
      propertyId,
      pricePerDay: '',
      name: '',
      dateRange: { from: null, to: null },
    },
  });
  const {
    reset,
    watch,
    handleSubmit,
    formState: { errors },
  } = methods;

  useEffect(() => {
    if (bookingId) {
      dispatch(fetchBookingById(bookingId)).then((response) => {
        const bookingData = response.payload;
        if (bookingData) {
          const formattedData = {
            ...bookingData,
            dateRange: {
              from: parseISO(bookingData.startDate),
              to: parseISO(bookingData.endDate),
            },
          };
          reset(formattedData);
        } else {
          setFormError('Failed to fetch booking data');
        }
      });
    }
  }, [bookingId, dispatch, reset]);

  const calculateTotalPrice = (
    startDate: Date,
    endDate: Date,
    pricePerDay: string,
  ) => {
    const diffTime = Math.abs(endDate.getTime() - startDate.getTime());
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24)) + 1;
    return diffDays * Number(pricePerDay);
  };

  useEffect(() => {
    const subscription = watch(({ dateRange, pricePerDay }) => {
      if (dateRange && dateRange.from && dateRange.to && pricePerDay) {
        setTotalPrice(
          calculateTotalPrice(dateRange.from, dateRange.to, pricePerDay),
        );
      }
    });
    return () => subscription.unsubscribe();
  }, [watch]);

  const isDateOverlap = (newStartDate: Date, newEndDate: Date) => {
    return bookings.some((booking) => {
      if (booking.propertyId !== propertyId) return false;
      if (bookingId && booking.id === bookingId) return false;

      const existingStartDate = format(
        startOfDay(parseISO(booking.startDate)),
        'yyyy-MM-dd',
      );
      const existingEndDate = format(
        startOfDay(parseISO(booking.endDate)),
        'yyyy-MM-dd',
      );
      const newStart = format(startOfDay(newStartDate), 'yyyy-MM-dd');
      const newEnd = format(startOfDay(newEndDate), 'yyyy-MM-dd');

      return (
        (newStart >= existingStartDate && newStart <= existingEndDate) ||
        (newEnd >= existingStartDate && newEnd <= existingEndDate) ||
        (existingStartDate >= newStart && existingStartDate <= newEnd)
      );
    });
  };

  const onSubmit = async (data: any) => {
    setIsLoading(true);
    const { dateRange, ...rest } = data;
    const newStartDate = startOfDay(dateRange.from);
    const newEndDate = startOfDay(dateRange.to);

    if (isDateOverlap(newStartDate, newEndDate)) {
      setFormError('The selected dates overlap with an existing booking.');
      setIsLoading(false);
      return;
    }

    const formattedStartDate = format(newStartDate, 'yyyy-MM-dd');
    const formattedEndDate = format(newEndDate, 'yyyy-MM-dd');

    const formattedData = {
      ...rest,
      startDate: formattedStartDate,
      endDate: formattedEndDate,
    };

    try {
      if (bookingId) {
        await dispatch(
          updateBooking({ bookingId, data: formattedData }),
        ).unwrap();
        toast({
          title: 'Booking updated successfully',
        });
      } else {
        await dispatch(createBooking(formattedData)).unwrap();
        toast({
          title: 'Booking created successfully',
        });
      }
      dispatch(fetchBookings(propertyId!));
      onClose();
    } catch (error: any) {
      setFormError(error.message || 'An error occurred');
      console.error('Error saving booking:', error);
    } finally {
      setIsLoading(false);
    }
  };

  const onDelete = async () => {
    setIsLoading(true);
    try {
      await dispatch(deleteBooking(bookingId!)).unwrap();
      toast({
        title: 'Booking deleted successfully',
      });
      dispatch(fetchBookings(propertyId!));
      onClose();
    } catch (error: any) {
      setFormError(error.message || 'An error occurred');
      console.error('Error deleting booking:', error);
    } finally {
      setIsLoading(false);
    }
  };

  if (loading) {
    return (
      <div className="h-[376px] content-center pb-12">
        <div className="flex justify-center">
          <ThreeDots
            visible={true}
            height="80"
            width="80"
            color="black"
            radius="9"
            ariaLabel="three-dots-loading"
          />
        </div>
      </div>
    );
  }

  return (
    <div className="flex min-h-full flex-1 flex-col justify-center">
      <div className="sm:mx-auto sm:w-full sm:max-w-[480px]">
        <div className="px-6">
          <FormProvider {...methods}>
            <form onSubmit={handleSubmit(onSubmit)} className="space-y-6">
              <FormField
                id="name"
                label="Guest Name"
                type="text"
                register={methods.register}
                errors={errors}
              />
              <div>
                <label
                  htmlFor="dateRange"
                  className="block text-sm font-medium leading-6 text-gray-900"
                >
                  Date Range
                </label>
                <div className="mt-2">
                  <DatePickerWithRange
                    name="dateRange"
                    currentBookingId={bookingId}
                  />
                  {errors.dateRange && (
                    <p className="text-red-500 text-sm mt-1">
                      {getErrorMessage(errors.dateRange)}
                    </p>
                  )}
                </div>
              </div>
              <FormField
                id="pricePerDay"
                label="Price per Day"
                type="number"
                register={methods.register}
                errors={errors}
              />
              <div>
                <div className="block text-sm font-medium leading-6 text-gray-900">
                  Total Price: {totalPrice.toFixed(2)}
                </div>
              </div>
              <div>
                {formError.length > 0 && (
                  <p className="text-red-500 text-sm mt-1">{formError}</p>
                )}
              </div>
              <div className="flex justify-between">
                {bookingId && (
                  <Button
                    type="button"
                    variant="destructive"
                    size="icon"
                    className="mr-3"
                    disabled={isLoading}
                    onClick={onDelete}
                  >
                    <TrashIcon className="h-5 w-5 text-white" />
                  </Button>
                )}
                <Button type="submit" className="w-full" disabled={isLoading}>
                  {bookingId ? 'Save Changes' : 'Create Booking'}
                </Button>
              </div>
            </form>
          </FormProvider>
        </div>
      </div>
    </div>
  );
}
