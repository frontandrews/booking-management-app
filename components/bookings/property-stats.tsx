import { DollarSign, ChevronLeft, ChevronRight, Users } from 'lucide-react';
import { Card, CardContent, CardHeader, CardTitle } from '../ui/card';
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '@/redux/store';
import {
  differenceInDays,
  endOfDay,
  startOfDay,
  startOfMonth,
  startOfYear,
  subDays,
  subMonths,
  subYears,
  format,
  addDays,
  addMonths,
  min,
  max,
  addYears,
} from 'date-fns';
import { Booking } from './types';

export function PropertyStats() {
  const bookings = useSelector(
    (state: RootState) => state.booking.bookings,
  ) as Booking[];
  const propertyId = useSelector(
    (state: RootState) => state.property.selectedPropertyId,
  ) as number;

  const [thisMonth, setThisMonth] = useState<number>(0);
  const [lastMonth, setLastMonth] = useState<number>(0);
  const [thisYear, setThisYear] = useState<number>(0);
  const [lastYear, setLastYear] = useState<number>(0);
  const [selectedMonth, setSelectedMonth] = useState<Date>(new Date());
  const [selectedMonthRevenue, setSelectedMonthRevenue] = useState<number>(0);
  const [previousMonthRevenue, setPreviousMonthRevenue] = useState<number>(0);
  const [guestsThisMonth, setGuestsThisMonth] = useState<number>(0);

  useEffect(() => {
    if (propertyId && bookings.length > 0) {
      const now = new Date();

      const startOfThisMonth = startOfMonth(now);
      const endOfThisMonth = endOfDay(
        addDays(startOfMonth(addMonths(now, 1)), -1),
      );
      const startOfLastMonth = startOfMonth(subMonths(now, 1));
      const endOfLastMonth = endOfDay(subDays(startOfThisMonth, 1));
      const startOfThisYear = startOfYear(now);
      const endOfThisYear = endOfDay(
        addDays(startOfYear(addYears(now, 1)), -1),
      );
      const startOfLastYear = startOfYear(subYears(now, 1));
      const endOfLastYear = endOfDay(subDays(startOfThisYear, 1));

      const thisMonthRevenue = calculateRevenue(
        startOfThisMonth,
        endOfThisMonth,
        propertyId,
        bookings,
      );
      const lastMonthRevenue = calculateRevenue(
        startOfLastMonth,
        endOfLastMonth,
        propertyId,
        bookings,
      );
      const thisYearRevenue = calculateRevenue(
        startOfThisYear,
        endOfThisYear,
        propertyId,
        bookings,
      );
      const lastYearRevenue = calculateRevenue(
        startOfLastYear,
        endOfLastYear,
        propertyId,
        bookings,
      );
      const selectedMonthStart = startOfMonth(selectedMonth);
      const selectedMonthEnd = endOfDay(
        addDays(startOfMonth(addMonths(selectedMonth, 1)), -1),
      );
      const selectedMonthRevenue = calculateRevenue(
        selectedMonthStart,
        selectedMonthEnd,
        propertyId,
        bookings,
      );

      const previousMonthStart = startOfMonth(subMonths(selectedMonth, 1));
      const previousMonthEnd = endOfDay(
        addDays(startOfMonth(selectedMonth), -1),
      );
      const previousMonthRevenue = calculateRevenue(
        previousMonthStart,
        previousMonthEnd,
        propertyId,
        bookings,
      );

      const guestsThisMonth = calculateGuestsThisMonth(
        startOfThisMonth,
        endOfThisMonth,
        propertyId,
        bookings,
      );

      setThisMonth(thisMonthRevenue);
      setLastMonth(lastMonthRevenue);
      setThisYear(thisYearRevenue);
      setLastYear(lastYearRevenue);
      setSelectedMonthRevenue(selectedMonthRevenue);
      setPreviousMonthRevenue(previousMonthRevenue);
      setGuestsThisMonth(guestsThisMonth);
    }
  }, [propertyId, bookings, selectedMonth]);

  const calculateRevenue = (
    startDate: Date,
    endDate: Date,
    propertyId: number,
    bookings: Booking[],
  ): number => {
    return bookings
      .filter(
        (booking) =>
          booking.propertyId === propertyId &&
          startOfDay(new Date(booking.startDate)) <= endDate &&
          endOfDay(new Date(booking.endDate)) >= startDate,
      )
      .reduce((total, booking) => {
        const bookingStart = max([
          startOfDay(new Date(booking.startDate)),
          startDate,
        ]);
        const bookingEnd = min([endOfDay(new Date(booking.endDate)), endDate]);
        const daysWithinRange = differenceInDays(bookingEnd, bookingStart) + 1;
        return total + daysWithinRange * Number(booking.pricePerDay);
      }, 0);
  };

  const calculateGuestsThisMonth = (
    startDate: Date,
    endDate: Date,
    propertyId: number,
    bookings: Booking[],
  ): number => {
    const guests = new Set(
      bookings.filter(
        (booking) =>
          booking.propertyId === propertyId &&
          startOfDay(new Date(booking.startDate)) <= endDate &&
          endOfDay(new Date(booking.endDate)) >= startDate,
      ),
    );
    return guests.size;
  };

  const handlePreviousMonth = () => {
    setSelectedMonth((prev) => subMonths(prev, 1));
  };

  const handleNextMonth = () => {
    setSelectedMonth((prev) => addMonths(prev, 1));
  };

  const formatCurrency = (value: number) => {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(value);
  };

  const formatPercentage = (current: number, previous: number) => {
    if (current === 0 && previous === 0) {
      return '0%';
    }
    if (previous === 0) {
      return '100%';
    }
    return `${(((current - previous) / previous) * 100).toFixed(2)}%`;
  };

  return (
    <div className="grid gap-4 md:grid-cols-2 md:gap-8 lg:grid-cols-4 pt-6">
      <Card className="rounded-none">
        <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
          <CardTitle className="text-sm font-medium">
            Guests This Month
          </CardTitle>
          <Users className="h-4 w-4 text-muted-foreground" />
        </CardHeader>
        <CardContent>
          <div className="text-2xl font-bold">{guestsThisMonth}</div>
        </CardContent>
      </Card>
      <Card className="rounded-none">
        <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
          <CardTitle className="text-sm font-medium">This Month</CardTitle>
          <DollarSign className="h-4 w-4 text-muted-foreground" />
        </CardHeader>
        <CardContent>
          <div className="text-2xl font-bold">{formatCurrency(thisMonth)}</div>
          <p className="text-xs text-muted-foreground">
            {formatPercentage(thisMonth, lastMonth)} from last month
          </p>
        </CardContent>
      </Card>
      <Card className="rounded-none">
        <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
          <CardTitle className="text-sm font-medium">This Year</CardTitle>
          <DollarSign className="h-4 w-4 text-muted-foreground" />
        </CardHeader>
        <CardContent>
          <div className="text-2xl font-bold">{formatCurrency(thisYear)}</div>
          <p className="text-xs text-muted-foreground">
            {formatPercentage(thisYear, lastYear)} from last year
          </p>
        </CardContent>
      </Card>
      <Card className="rounded-none">
        <CardHeader className="flex flex-row items-center justify-between space-y-0 pb-2">
          <CardTitle className="text-sm font-medium flex">
            <button onClick={handlePreviousMonth}>
              <ChevronLeft className="h-4 w-4" />
            </button>
            <div className="min-w-[110px] text-center">
              {format(selectedMonth, 'MMMM yyyy')}
            </div>
            <button onClick={handleNextMonth}>
              <ChevronRight className="h-4 w-4" />
            </button>
          </CardTitle>
          <DollarSign className="h-4 w-4 text-muted-foreground" />
        </CardHeader>
        <CardContent>
          <div className="text-2xl font-bold">
            {formatCurrency(selectedMonthRevenue)}
          </div>
          <p className="text-xs text-muted-foreground">
            {formatPercentage(selectedMonthRevenue, previousMonthRevenue)} from
            last month
          </p>
        </CardContent>
      </Card>
    </div>
  );
}
