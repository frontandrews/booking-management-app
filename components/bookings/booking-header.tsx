import React from 'react';
import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/react/20/solid';
import { format } from 'date-fns';
import { BookingDialog } from './booking-dialog';

interface BookingHeaderProps {
  currentMonth: Date;
  onPreviousMonth: () => void;
  onNextMonth: () => void;
  onToday: () => void;
}

export const BookingHeader = ({
  currentMonth,
  onPreviousMonth,
  onNextMonth,
  onToday,
}: BookingHeaderProps) => {
  return (
    <header className="flex flex-col items-center justify-between border-b border-gray-200 py-4 lg:flex-row lg:flex-none">
      <h1 className="text-2xl font-semibold leading-6 text-gray-900 mb-4 lg:mb-0">
        <time dateTime={format(currentMonth, 'yyyy-MM')}>
          {format(currentMonth, 'MMMM yyyy')}
        </time>
      </h1>
      <div className="flex flex-col items-center lg:flex-row lg:space-x-4">
        <div className="relative flex items-center rounded-md bg-white shadow-sm mb-4 lg:mb-0 md:items-stretch">
          <button
            type="button"
            className="flex h-9 w-12 items-center justify-center rounded-l-md border-y border-l border-gray-300 pr-1 text-gray-400 hover:text-gray-500 focus:relative md:w-9 md:pr-0 md:hover:bg-gray-50"
            onClick={onPreviousMonth}
          >
            <span className="sr-only">Previous month</span>
            <ChevronLeftIcon className="h-5 w-5" aria-hidden="true" />
          </button>
          <button
            type="button"
            className="border-x flex h-9 items-center justify-center border-y border-gray-300 px-3.5 text-sm font-semibold text-gray-900 hover:bg-gray-50 focus:relative md:block"
            onClick={onToday}
          >
            Today
          </button>
          <span className="relative -mx-px h-9 w-px bg-gray-300 md:hidden" />
          <button
            type="button"
            className="flex h-9 w-12 items-center justify-center rounded-r-md border-y border-r border-gray-300 pl-1 text-gray-400 hover:text-gray-500 focus:relative md:w-9 md:pl-0 md:hover:bg-gray-50"
            onClick={onNextMonth}
          >
            <span className="sr-only">Next month</span>
            <ChevronRightIcon className="h-5 w-5" aria-hidden="true" />
          </button>
        </div>
        <BookingDialog />
      </div>
    </header>
  );
};
