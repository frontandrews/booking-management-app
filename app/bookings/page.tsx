'use client';

import { useEffect, useRef, useState } from 'react';
import { PropertyDialog } from '@/components/bookings/property-dialog';
import PropertyList from '@/components/bookings/property-list';
import SectionHeading from '@/components/ui/section-heading';
import BookingView from '@/components/bookings/bookings-view';
import {
  fetchProperties,
  setSelectedPropertyId,
} from '@/redux/features/property/slice';
import EmptyState from '@/components/ui/empty-state';
import { ThreeDots } from 'react-loader-spinner';
import { fetchBookings } from '@/redux/features/booking/slice';
import { PropertyStats } from '@/components/bookings/property-stats';
import { useAppDispatch, useAppSelector } from '@/redux/hooks';

export default function BookingsPage() {
  const dispatch = useAppDispatch();
  const selectedPropertyId = useAppSelector(
    (state) => state.property.selectedPropertyId,
  );
  const bookings = useAppSelector((state) => state.booking.bookings);
  const properties = useAppSelector((state) => state.property.properties);
  const loading = useAppSelector((state) => state.property.loading);

  const [initialLoading, setInitialLoading] = useState(true);
  const hasAutoSelected = useRef(false);

  useEffect(() => {
    if (properties.length === 0) {
      dispatch(fetchProperties()).then(() => setInitialLoading(false));
    } else {
      setInitialLoading(false);
    }
  }, [properties.length, dispatch]);

  // Auto select first property
  useEffect(() => {
    if (!hasAutoSelected.current && properties.length > 0) {
      const id = properties[0].id;
      if (id) {
        if (id !== selectedPropertyId) {
          dispatch(setSelectedPropertyId(id));
        }
        dispatch(fetchBookings(id));
      }
      hasAutoSelected.current = true;
    }
  }, [properties, selectedPropertyId, dispatch]);

  return (
    <div className="container mx-auto">
      <SectionHeading title="Dashboard" />

      {initialLoading || loading ? (
        <div className="flex justify-center py-10">
          <ThreeDots
            visible={true}
            height="80"
            width="80"
            color="black"
            radius="9"
            ariaLabel="three-dots-loading"
          />
        </div>
      ) : properties.length === 0 ? (
        <EmptyState
          title="No properties found"
          description="Create a new property to get started."
        >
          <PropertyDialog />
        </EmptyState>
      ) : (
        <>
          <div className="grid grid-cols-12 gap-4">
            <div className="col-span-12 lg:col-span-4">
              <PropertyList />
            </div>
            <div className="col-span-12 lg:col-span-8">
              <BookingView bookings={bookings} />
            </div>
          </div>
          <PropertyStats />
        </>
      )}
    </div>
  );
}
