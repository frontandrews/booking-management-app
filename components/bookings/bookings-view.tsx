import React, { useState, useEffect } from 'react';
import {
  startOfMonth,
  endOfMonth,
  startOfWeek,
  endOfWeek,
  addDays,
  addMonths,
  subMonths,
} from 'date-fns';
import { BookingHeader } from './booking-header';
import { Calendar } from './calendar';
import { Day, BookingViewProps } from './types';

const BookingView = ({ bookings = [] }: BookingViewProps) => {
  const [currentMonth, setCurrentMonth] = useState<Date>(new Date());
  const [days, setDays] = useState<Day[]>([]);

  useEffect(() => {
    const generateDays = (date: Date): Day[] => {
      const startDate = startOfWeek(startOfMonth(date));
      const endDate = endOfWeek(endOfMonth(date));
      const days: Day[] = [];
      let day = startDate;

      while (day <= endDate) {
        days.push(day);
        day = addDays(day, 1);
      }
      return days;
    };
    setDays(generateDays(currentMonth));
  }, [currentMonth]);

  const handlePreviousMonth = () => {
    setCurrentMonth(subMonths(currentMonth, 1));
  };

  const handleNextMonth = () => {
    setCurrentMonth(addMonths(currentMonth, 1));
  };

  const handleToday = () => {
    setCurrentMonth(new Date());
  };

  return (
    <div className="lg:flex lg:h-full lg:flex-col px-4 pb-4 border">
      <BookingHeader
        currentMonth={currentMonth}
        onPreviousMonth={handlePreviousMonth}
        onNextMonth={handleNextMonth}
        onToday={handleToday}
      />
      <Calendar days={days} currentMonth={currentMonth} bookings={bookings} />
    </div>
  );
};

export default BookingView;
