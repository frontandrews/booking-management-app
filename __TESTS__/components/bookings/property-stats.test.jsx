import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { PropertyStats } from '@/components/bookings/property-stats';

const createMockStore = (initialState) => {
  return configureStore({
    reducer: {
      booking: () => initialState.booking,
      property: () => initialState.property,
    },
    preloadedState: initialState,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  });
};

describe('PropertyStats Component', () => {
  let store;

  const initialBookingState = {
    bookings: [
      {
        id: 1,
        propertyId: 1,
        startDate: '2023-06-01',
        endDate: '2023-06-10',
        pricePerDay: '100',
      },
      {
        id: 2,
        propertyId: 1,
        startDate: '2023-06-15',
        endDate: '2023-06-20',
        pricePerDay: '150',
      },
      {
        id: 3,
        propertyId: 1,
        startDate: '2023-07-01',
        endDate: '2023-07-05',
        pricePerDay: '200',
      },
    ],
  };

  const initialPropertyState = {
    properties: [
      { id: 1, name: 'Property One', location: 'Location One' },
      { id: 2, name: 'Property Two', location: 'Location Two' },
    ],
    selectedPropertyId: 1,
    loading: false,
    error: null,
  };

  const initialState = {
    booking: initialBookingState,
    property: initialPropertyState,
  };

  beforeEach(() => {
    store = createMockStore(initialState);
  });

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <PropertyStats />
      </Provider>,
    );

  test('renders correctly and displays the initial stats', () => {
    renderComponent();

    expect(screen.getByText('Guests This Month')).toBeInTheDocument();
    expect(screen.getByText('This Month')).toBeInTheDocument();
    expect(screen.getByText('This Year')).toBeInTheDocument();
  });
});
