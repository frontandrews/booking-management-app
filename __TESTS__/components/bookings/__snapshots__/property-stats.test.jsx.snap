// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyStats Component renders correctly and displays the initial stats 1`] = `
<div>
  <div
    class="grid gap-4 md:grid-cols-2 md:gap-8 lg:grid-cols-4 pt-6"
  >
    <div
      class="border bg-card text-card-foreground shadow-sm rounded-none"
    >
      <div
        class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
      >
        <h3
          class="tracking-tight text-sm font-medium"
        >
          Guests This Month
        </h3>
        <svg
          class="lucide lucide-users h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
          <path
            d="M22 21v-2a4 4 0 0 0-3-3.87"
          />
          <path
            d="M16 3.13a4 4 0 0 1 0 7.75"
          />
        </svg>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class="text-2xl font-bold"
        >
          0
        </div>
      </div>
    </div>
    <div
      class="border bg-card text-card-foreground shadow-sm rounded-none"
    >
      <div
        class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
      >
        <h3
          class="tracking-tight text-sm font-medium"
        >
          This Month
        </h3>
        <svg
          class="lucide lucide-dollar-sign h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="22"
          />
          <path
            d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
          />
        </svg>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class="text-2xl font-bold"
        >
          $0
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          0%
           from last month
        </p>
      </div>
    </div>
    <div
      class="border bg-card text-card-foreground shadow-sm rounded-none"
    >
      <div
        class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
      >
        <h3
          class="tracking-tight text-sm font-medium"
        >
          This Year
        </h3>
        <svg
          class="lucide lucide-dollar-sign h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="22"
          />
          <path
            d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
          />
        </svg>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class="text-2xl font-bold"
        >
          $0
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          -100.00%
           from last year
        </p>
      </div>
    </div>
    <div
      class="border bg-card text-card-foreground shadow-sm rounded-none"
    >
      <div
        class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
      >
        <h3
          class="tracking-tight text-sm font-medium flex"
        >
          <button
            aria-label="Previous month"
          >
            <svg
              class="lucide lucide-chevron-left h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15 18-6-6 6-6"
              />
            </svg>
          </button>
          <div
            class="min-w-[110px] text-center"
          >
            July 2024
          </div>
          <button
            aria-label="Next month"
          >
            <svg
              class="lucide lucide-chevron-right h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
          </button>
        </h3>
        <svg
          class="lucide lucide-dollar-sign h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="22"
          />
          <path
            d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
          />
        </svg>
      </div>
      <div
        class="p-6 pt-0"
      >
        <div
          class="text-2xl font-bold"
        >
          $0
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          0%
           from last month
        </p>
      </div>
    </div>
  </div>
</div>
`;
